File: src/db/class_table.py
from sqlalchemy import Column, Integer, String
from src.definitions import Base
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Sequence
from sqlalchemy import select


class ClassTable(Base):
    """
    Example:

    An example instance of ClassTable in JSON format would look like this:

    ```json
        {
            "flurry_of_blows": "None",
            "slots_1": "None",
            "slots_8": "None",
            "spells_known_5": "None",
            "name": "Barbarian",
            "bonus_spells": "None",
            "slots_2": "None",
            "slots_9": "None",
            "spells_known_7": "None",
            "fort_save": "+7",
            "powers_known": "None",
            "slots_3": "None",
            "spells_known_0": "None",
            "spells_known_8": "None",
            "ref_save": "+3",
            "unarmored_speed_bonus": "None",
            "slots_4": "None",
            "spells_known_1": "None",
            "spells_known_9": "None",
            "will_save": "+3",
            "unarmed_damage": "None",
            "slots_5": "None",
            "spells_known_2": "None",
            "reference": "SRD 3.5 ClassesI",
            "caster_level": "None",
            "power_level": "None",
            "slots_6": "None",
            "spells_known_3": "None",
            "level": "11",
            "points_per_day": "None",
            "special": "Greater rage",
            "slots_7": "None",
            "spells_known_4": "None",
            "id": 11,
            "base_attack_bonus": "+11/+6/+1",
            "ac_bonus": "None",
            "slots_0": "None",
            "spells_known_6": "None"
        },
    ```
    """

    __tablename__ = "class_table"
    id = Column("id", Integer, primary_key=True)
    name = Column("name", String)
    level = Column("level", String)
    base_attack_bonus = Column("base_attack_bonus", String)
    fort_save = Column("fort_save", String)
    ref_save = Column("ref_save", String)
    will_save = Column("will_save", String)
    caster_level = Column("caster_level", String)
    points_per_day = Column("points_per_day", String)
    ac_bonus = Column("ac_bonus", String)
    flurry_of_blows = Column("flurry_of_blows", String)
    bonus_spells = Column("bonus_spells", String)
    powers_known = Column("powers_known", String)
    unarmored_speed_bonus = Column("unarmored_speed_bonus", String)
    unarmed_damage = Column("unarmed_damage", String)
    power_level = Column("power_level", String)
    special = Column("special", String)
    slots_0 = Column("slots_0", String)
    slots_1 = Column("slots_1", String)
    slots_2 = Column("slots_2", String)
    slots_3 = Column("slots_3", String)
    slots_4 = Column("slots_4", String)
    slots_5 = Column("slots_5", String)
    slots_6 = Column("slots_6", String)
    slots_7 = Column("slots_7", String)
    slots_8 = Column("slots_8", String)
    slots_9 = Column("slots_9", String)
    spells_known_0 = Column("spells_known_0", String)
    spells_known_1 = Column("spells_known_1", String)
    spells_known_2 = Column("spells_known_2", String)
    spells_known_3 = Column("spells_known_3", String)
    spells_known_4 = Column("spells_known_4", String)
    spells_known_5 = Column("spells_known_5", String)
    spells_known_6 = Column("spells_known_6", String)
    spells_known_7 = Column("spells_known_7", String)
    spells_known_8 = Column("spells_known_8", String)
    spells_known_9 = Column("spells_known_9", String)
    reference = Column("reference", String)


class ClassTableRepository:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def get_all(self, page: int = 1, page_size: int = 10) -> Sequence[ClassTable]:
        query = select(ClassTable).offset(page_size * page).limit(page_size)
        return (await self.db.execute(query)).scalars().all()

File: src/db/class_table.go
package db

import (
	"database/sql"
	"fmt"
	"reflect"

	"github.com/Masterminds/squirrel"
)

type ClassTableColumns struct {
	ID                  int            `db:"id"`
	Name                string         `db:"name"`
	Level               sql.NullString `db:"level"`
	BaseAttackBonus     sql.NullString `db:"base_attack_bonus"`
	FortSave            sql.NullString `db:"fort_save"`
	RefSave             sql.NullString `db:"ref_save"`
	WillSave            sql.NullString `db:"will_save"`
	CasterLevel         sql.NullString `db:"caster_level"`
	PointsPerDay        sql.NullString `db:"points_per_day"`
	AcBonus             sql.NullString `db:"ac_bonus"`
	FlurryOfBlows       sql.NullString `db:"flurry_of_blows"`
	BonusSpells         sql.NullString `db:"bonus_spells"`
	PowersKnown         sql.NullString `db:"powers_known"`
	UnarmoredSpeedBonus sql.NullString `db:"unarmored_speed_bonus"`
	UnarmedDamage       sql.NullString `db:"unarmed_damage"`
	PowerLevel          sql.NullString `db:"power_level"`
	Special             sql.NullString `db:"special"`
	Slots0              sql.NullString `db:"slots_0"`
	Slots1              sql.NullString `db:"slots_1"`
	Slots2              sql.NullString `db:"slots_2"`
	Slots3              sql.NullString `db:"slots_3"`
	Slots4              sql.NullString `db:"slots_4"`
	Slots5              sql.NullString `db:"slots_5"`
	Slots6              sql.NullString `db:"slots_6"`
	Slots7              sql.NullString `db:"slots_7"`
	Slots8              sql.NullString `db:"slots_8"`
	Slots9              sql.NullString `db:"slots_9"`
	SpellsKnown0        sql.NullString `db:"spells_known_0"`
	SpellsKnown1        sql.NullString `db:"spells_known_1"`
	SpellsKnown2        sql.NullString `db:"spells_known_2"`
	SpellsKnown3        sql.NullString `db:"spells_known_3"`
	SpellsKnown4        sql.NullString `db:"spells_known_4"`
	SpellsKnown5        sql.NullString `db:"spells_known_5"`
	SpellsKnown6        sql.NullString `db:"spells_known_6"`
	SpellsKnown7        sql.NullString `db:"spells_known_7"`
	SpellsKnown8        sql.NullString `db:"spells_known_8"`
	SpellsKnown9        sql.NullString `db:"spells_known_9"`
	Reference           sql.NullString `db:"reference"`
}

var ClassTableTableName = "class_table"

type ClassTableRepo struct {
	*BaseRepo
	Class ClassListColumns
}

func (ClassTableRepo) FindAll(page uint64, pageSize uint64) (*sql.Rows, error) {
	sql, args, err := squirrel.Select("*").
		From(ClassTableTableName).
		Offset(page * pageSize).
		Limit(pageSize).
		ToSql()
	if err != nil {
		return nil, err
	}
	rows, err := DB.Query(sql, args...)
	if err != nil {
		return nil, err
	}
	return rows, nil
}

func (c ClassTableRepo) ExtractVars() map[string]interface{} {
	v := reflect.ValueOf(c.Class)
	t := v.Type()
	var fields = make(map[string]interface{})
	for i := 0; i < v.NumField(); i++ {
		field := v.Field(i)
		tag := t.Field(i).Tag.Get("db")
		switch field.Interface().(type) {
		case sql.NullString:
			ns := field.Interface().(sql.NullString)
			if ns.Valid {
				fields[tag] = ns.String
			}
		case string:
			if field.String() != "" {
				fields[tag] = field.String()
			}
		case int:
			if field.Int() != 0 {
				fields[tag] = fmt.Sprintf("%d", field.Int())
			}
		}
	}
	return fields
}

func (c ClassTableRepo) Find(filter ClassTableColumns, page uint64, pageSize uint64) ClassTableRepo {
	c.QueryBuilder = squirrel.Select("*").From(ClassTableTableName)
	return c
}
